File: src/testing/testing.test.ts
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { delimiter, join } from "node:path";
import { test } from "node:test";

import { expect } from "expect";

import TempDir from "../TempDir.ts";
import { loadConfig } from "../config.ts";
import { InMemoryPackageSource, type SourceName } from "../packages/index.ts";
import {
  compileTests,
  compileTestsArgs,
  CONSOLE_LAUNCHER,
  consoleLauncherJar,
  findTestSources,
  mainClassesDir,
  testClassesDir,
  testRunArgs,
} from "./testing.ts";

// the launcher download lands in the package store: isolate it
const STORE = TempDir.create("cappu-test-store-").path;
process.env.CAPPU_PACKAGE_STORE = STORE;

function tempProject(): TempDir {
  return TempDir.create("cappu-testing-");
}

test("test sources come from src/test/java; a missing dir means none", () => {
  using project = tempProject();
  const dir = project.path;
  const config = loadConfig(undefined, dir);
  expect(findTestSources(config)).toEqual([]);
  mkdirSync(join(dir, "src", "test", "java"), { recursive: true });
  writeFileSync(join(dir, "src", "test", "java", "ATest.java"), "class ATest {}");
  expect(findTestSources(config)).toEqual([join(dir, "src", "test", "java", "ATest.java")]);
});

test("test compile and run classpaths are ordered and jar-expanded", () => {
  using project = tempProject();
  const dir = project.path;
  mkdirSync(join(dir, ".cappu", "lib", "classes"), { recursive: true });
  mkdirSync(join(dir, ".cappu", "lib", "test-classes"), { recursive: true });
  writeFileSync(join(dir, ".cappu", "lib", "classes", "dep.jar"), "x");
  writeFileSync(join(dir, ".cappu", "lib", "test-classes", "junit.jar"), "x");
  const config = loadConfig(undefined, dir);

  const args = compileTestsArgs(config, ["/t/ATest.java"]);
  const cp = args[args.indexOf("-cp") + 1]!;
  expect(args[args.indexOf("-d") + 1]).toBe(join(dir, ".cappu", "test-build", "test-classes"));
  expect(cp.split(delimiter)).toEqual([
    mainClassesDir(config),
    join(dir, ".cappu", "lib", "classes"),
    join(dir, ".cappu", "lib", "classes", "dep.jar"),
    join(dir, ".cappu", "lib", "test-classes"),
    join(dir, ".cappu", "lib", "test-classes", "junit.jar"),
  ]);

  const run = testRunArgs(config, "/store/launcher.jar");
  expect(run.slice(0, 3)).toEqual(["-jar", "/store/launcher.jar", "execute"]);
  const runCp = run[run.indexOf("--class-path") + 1]!;
  expect(runCp.split(delimiter)[0]).toBe(join(dir, ".cappu", "test-build", "test-classes"));
  expect(run.at(-1)).toBe("--scan-class-path");
  // default outputFormat is "text": no report is written
  expect(run).not.toContain("--reports-dir");
});

test("testRunArgs writes junit reports to the resolved reportsDir only for junit", () => {
  using project = tempProject();
  const dir = project.path;
  const config = loadConfig(undefined, dir);

  config.testOptions.outputFormat = "junit";
  const run = testRunArgs(config, "/store/launcher.jar");
  expect(run[run.indexOf("--reports-dir") + 1]).toBe(join(dir, "dist", "test-results"));
  expect(run.at(-1)).toBe("--scan-class-path"); // still last

  config.testOptions.reportsDir = "./build/reports";
  const custom = testRunArgs(config, "/store/launcher.jar");
  expect(custom[custom.indexOf("--reports-dir") + 1]).toBe(join(dir, "build", "reports"));
});

test("compileTests wipes stale class files first (no phantom tests)", () => {
  using project = tempProject();
  const dir = project.path;
  const config = loadConfig(undefined, dir);
  const classes = testClassesDir(config);
  mkdirSync(join(classes, "old"), { recursive: true });
  writeFileSync(join(classes, "old", "GoneTest.class"), "stale");
  // a no-op javac (status 0) still triggers the pre-compile wipe
  const diagnostics = compileTests(config, ["/t/ATest.java"], () => ({ status: 0, stderr: "" }));
  expect(diagnostics).toEqual([]);
  expect(existsSync(join(classes, "old", "GoneTest.class"))).toBe(false);
});

test("test-compile failures map javac stderr to diagnostics", () => {
  using project = tempProject();
  const dir = project.path;
  const config = loadConfig(undefined, dir);
  const diagnostics = compileTests(config, ["/t/ATest.java"], () => ({
    status: 1,
    stderr: "/t/ATest.java:4: error: cannot find symbol\n1 error\n",
  }));
  expect(diagnostics).toEqual([
    { severity: "error", file: "/t/ATest.java", line: 4, message: "cannot find symbol" },
  ]);
  const missing = compileTests(config, ["/t/ATest.java"], () => ({
    status: null,
    stderr: "",
    error: new Error("ENOENT"),
  }));
  expect(missing[0]!.message).toContain("needs javac");
});

test("the console launcher is fetched once into the package store", async () => {
  using project = tempProject();
  const dir = project.path;
  const config = loadConfig(undefined, dir);
  let downloads = 0;
  const source = {
    name: "mem" as SourceName,
    search: () => Promise.resolve([]),
    listVersions: () => Promise.resolve([]),
    getMetadata: () => Promise.resolve(undefined),
    getArtifact: () => {
      downloads++;
      return Promise.resolve(new TextEncoder().encode("launcher-bytes"));
    },
  };

  const first = await consoleLauncherJar(config, [source]);
  expect(first).toContain(CONSOLE_LAUNCHER.artifactId);
  expect(first.startsWith(STORE)).toBe(true);
  const second = await consoleLauncherJar(config, [source]);
  expect(second).toBe(first);
  expect(downloads).toBe(1);

  await expect(consoleLauncherJar(config, [new InMemoryPackageSource("empty", [])])).resolves.toBe(
    first,
  ); // still cached even with a useless source
});

File: togo/internal/testing/testing_test.go
package testing

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/nikeee/cappu/internal/config"
)

func project(t *testing.T) *config.Config {
	t.Helper()
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, "cappu.json"), []byte("{}"), 0o644); err != nil {
		t.Fatal(err)
	}
	cfg, err := config.Load("", dir)
	if err != nil {
		t.Fatal(err)
	}
	return cfg
}

func TestFindTestSources(t *testing.T) {
	// A project with no src/test/java yields no test sources.
	if got := FindTestSources(project(t)); len(got) != 0 {
		t.Errorf("missing test dir should yield none, got %v", got)
	}

	cfg := project(t)
	src := filepath.Join(cfg.BaseDir, config.DefaultTestSourcePath, "com", "example")
	if err := os.MkdirAll(src, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(src, "T.java"), []byte("class T {}"), 0o644); err != nil {
		t.Fatal(err)
	}
	got := FindTestSources(cfg)
	if len(got) != 1 || !strings.HasSuffix(got[0], filepath.Join("com", "example", "T.java")) {
		t.Errorf("FindTestSources = %v", got)
	}
}

// CompileTests returns located diagnostics (not a flat error). When javac cannot
// run at all it collapses to one error diagnostic, matching compileTests in TS.
func TestCompileTestsJavacUnavailable(t *testing.T) {
	cfg := project(t)
	cfg.CompilerOptions.Javac = "cappu-no-such-javac"
	diags := CompileTests(cfg, []string{"X.java"})
	if len(diags) != 1 || diags[0].Severity != "error" || !strings.Contains(diags[0].Message, "compiling tests needs javac") {
		t.Errorf("diags = %+v", diags)
	}
}

func TestRunArgsStructure(t *testing.T) {
	cfg := project(t)
	args := TestRunArgs(cfg, "/path/launcher.jar")
	if args[0] != "-jar" || args[1] != "/path/launcher.jar" || args[2] != "execute" {
		t.Errorf("prefix = %v", args[:3])
	}
	joined := strings.Join(args, " ")
	if !strings.Contains(joined, "--class-path") {
		t.Errorf("missing class-path flag: %v", args)
	}
	// --scan-class-path must be the last argument (matches the TS run.at(-1)).
	if args[len(args)-1] != "--scan-class-path" {
		t.Errorf("last arg = %q, want --scan-class-path", args[len(args)-1])
	}
	// the compiled test classes must be the first runtime classpath entry
	if !strings.Contains(args[4], TestClassesDir(cfg)) {
		t.Errorf("test classes not on the runtime classpath: %q", args[4])
	}
	// default outputFormat is "text": no report is written
	if strings.Contains(strings.Join(args, " "), "--reports-dir") {
		t.Errorf("text format must not write reports: %v", args)
	}
}

func TestRunArgsJunitReports(t *testing.T) {
	cfg := project(t)
	cfg.TestOptions.OutputFormat = "junit"
	args := TestRunArgs(cfg, "/path/launcher.jar")
	i := indexOf(args, "--reports-dir")
	if i < 0 || args[i+1] != cfg.ResolvePath(config.DefaultTestReportsDir) {
		t.Errorf("--reports-dir = %v, want %q", args, cfg.ResolvePath(config.DefaultTestReportsDir))
	}
	// --scan-class-path stays last
	if args[len(args)-1] != "--scan-class-path" {
		t.Errorf("last arg = %q, want --scan-class-path", args[len(args)-1])
	}

	cfg.TestOptions.ReportsDir = "./build/reports"
	custom := TestRunArgs(cfg, "/path/launcher.jar")
	i = indexOf(custom, "--reports-dir")
	if i < 0 || custom[i+1] != cfg.ResolvePath("./build/reports") {
		t.Errorf("custom --reports-dir = %v", custom)
	}
}

func indexOf(s []string, v string) int {
	for i, x := range s {
		if x == v {
			return i
		}
	}
	return -1
}

func TestRuntimeClassPathOrder(t *testing.T) {
	cfg := project(t)
	cp := TestRuntimeClassPath(cfg)
	if cp[0] != TestClassesDir(cfg) {
		t.Errorf("runtime cp[0] = %q, want test-classes dir", cp[0])
	}
	if cp[1] != MainClassesDir(cfg) {
		t.Errorf("runtime cp[1] = %q, want main classes dir", cp[1])
	}
}
